master/pkg/searcher/asha_stopping.go

```diff
@@ -0,0 +1,259 @@
+package searcher
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/determined-ai/determined/master/pkg/workload"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AsyncHalvingStoppingSearch implements a modified version of the asynchronous successive
+// halving algorithm (ASHA) that does not require fault tolerance to perform early-stopping.
+// For each trial, after a train and validation workload, the algorithm will decide whether
+// to stop or continue training the trial based on the ranking of the validation metric
+// compared to other trials in a particular rung.  Once a trial has been stopped, it will not
+// be resumed later; this is why the algorithm does not require fault tolerance.
+// The searcher state and config match that of AsyncHalvingSearch but we will only run
+// the stopping based version if StopOnce is true.
+type asyncHalvingStoppingSearch struct {
+	defaultSearchMethod
+	model.AsyncHalvingConfig
+	asyncHalvingSearchState
+}
+
+func newAsyncHalvingStoppingSearch(config model.AsyncHalvingConfig) SearchMethod {
+	rungs := make([]*rung, 0, config.NumRungs)
+	if !config.StopOnce {
+		panic(fmt.Sprintf("stop_once is %t, should not be running stopping-based ASHA", config.StopOnce))
```

Comment:
this seems unnecessary but i'm fine with it if you think it will prevent misuse one day.

Reply:
took a look and agree that given how we create the searchers, this is unnecessary.